[core-https] core 2.0 background/writeup (#10169)
Azure/azure-sdk-for-js: sdk/core/core-https/documentation/core2.md

> # Azure SDK for JS Core 2.0

Reviewer: Azure Core is the term we have been using more often outside the JS team, so perhaps people may relate to the below better?

```suggestion
# Azure Core 2.0 for JS SDKs
```

## What do you mean 2.0?

> Today one of the primary core packages we ship is `@azure/core-http`. As of this document being written it sits at version 1.1.5. Due to a number factors discussed in this document, there is a desire to make some breaking changes that would create a new major version of this package (i.e. "2.0.0".)

Reviewer: A line talking about what core-http does would be good here.

```suggestion
Today, in Azure SDKs for JS, one of the primary core packages we ship is `@azure/core-http`. This package forms the runtime for all the auto-generated Azure SDKs by providing the mechanism to route HTTP requests through a pipeline made of different policies like retry, logging, tracing etc. As of this document being written it sits at version 1.1.5. Due to a number factors discussed in this document, there is a desire to create a new major version of this package (i.e. "2.0.0") that would include breaking changes.
```

> Since `@azure/core-http` is intended only for indirect use by most consumers, a major version change would not be largely disruptive. Currently, in order to preserve compatibility and not complicate the story for servicing, a new package has been created that is currently named `@azure/core-https`. This name is not final.

Reviewer: ```suggestion
Azure customers are indirect consumers of this package via the SDKs for different Azure services. Therefore, a major version change would not be largely disruptive. Currently, in order to preserve compatibility and not complicate the story for servicing, a new package has been created that is currently named `@azure/core-https`. This name is not final.
```

> ## Motivations, Context, and History

Reviewer: Since history comes first...

```suggestion
## History, Context, and Motivations
```

In ancient times (~5 years ago), Microsoft provided SDKs for NodeJS developers wanting to interact with Azure services. Ultimately, many of these packages were generated using [AutoRest](https://github.com/Azure/Autorest) using a package called `autorest.nodejs`. The autogenerated code had the need for a common, re-usable core for performing HTTP operations, and thus [`ms-rest`](https://github.com/Azure/ms-rest-js) was created.

> Besides `ms-rest` for common code, another package `ms-rest-azure` was created to house Azure-specific components and authetication code that worked with credentials. At this point, support for TypeScript came in the form of hand-authored d.ts files.

Reviewer: Here, does `common code` apply to `ms-rest` or `ms-rest-azure`? Me confused. How about skipping that part altogether? Extending the creds part with some examples will provide more clarity

```suggestion
Another package `ms-rest-azure` was created to house Azure-specific components and authetication code that worked with a variety of credentials like Service Principal, Device Token, Managed Identity etc. At this point, support for TypeScript came in the form of hand-authored type declaration files.
```

> Later, there emerged a desire for isomorphic packages, packages that worked in both the browser as well as Node. As part of this effort, code was to be generated in TypeScript directly and a new repo was created: `azure-sdk-for-js`.

Reviewer: ```suggestion
Later, there emerged a desire for isomorphic packages, packages that worked in both the browser as well as Node. This required a new runtime and code generator. As part of this effort, code was generated in TypeScript directly and a new repo was created: `azure-sdk-for-js`.
```

> A new code generator was created (`autorest.typescript`) that replaced the existing one (`autorest.nodejs`.) `ms-rest` became `@azure/ms-rest-js`. The authentication pieces from `@ms-rest-azure` were put into `@azure/ms-rest-nodeauth` and an equivalent package was created for the browser called `@azure/ms-rest-browserauth`. All remaining parts of `ms-rest-azure` were placed into `@azure/ms-rest-azure-js`.

Reviewer: ```suggestion
A new code generator was created (`autorest.typescript`) that replaced the existing one (`autorest.nodejs`.) `ms-rest` became `@azure/ms-rest-js`. The authentication pieces from `ms-rest-azure` were put into `@azure/ms-rest-nodeauth` and an equivalent package was created for the browser called `@azure/ms-rest-browserauth`. All remaining parts of `ms-rest-azure` were placed into `@azure/ms-rest-azure-js`.
```

> `@azure/ms-rest-js` was used as the basis for `@azure/core-http` when the JS team for the new Azure SDK was formed. This was nice as it bootstrapped the ability to make HTTP requests and provided broad compatibility with many track 1 clients that still exist in the `@azure/sdk-for-js` repository today.

Reviewer: Are we trying to say that `core-http` can be used with Track 1 clients?

Author: I think so? Don't we? Or do those only rely on `ms-rest-js`?

Reviewer: Track 1 clients are all auto generated using v4 of the code generator and use `@azure/ms-rest-js` and`@azure/ ms-rest-nodeauth`
If they use v5 of the code generator, then will be using `@azure/core-http` and `@azure/identity`

Author: Hmm so should we say that it was *originally* designed for interop, but we have since relaxed that as a requirement?

Reviewer: So, this section is about the benefits of using ms-rest-js as the basis for core-http.

The only benefit worth calling out is that the move from Track 1 to Track 2 for auto generated libraries is made easier by relying on core-http which was almost the same as ms-rest-js. The other one is that ms-rest-js was a tried and tested, and we had no reason at the time to create anything from scratch

> The downside is that much of the code had evolved organically over time, experienced incomplete upgrades to various layers, and had dead code that wasn't really needed anymore. We mitigated this during the 1.0 GA timeframe by manually declaring cleaner public interfaces (e.g. `OperationOptions` over `RequestOptionsBase`) but many of the internal interfaces were still messy, loosely typed, or confusing.

Reviewer: ```suggestion
The downside is that much of the code had evolved organically over time, experienced incomplete upgrades to various layers, and had dead code that wasn't really needed anymore. We mitigated this during the 1.0 GA timeframe in Nov 2020 by manually declaring cleaner public interfaces (e.g. `OperationOptions` over `RequestOptionsBase`) but many of the internal interfaces were still messy, loosely typed, or confusing.
```

As the JS team has gained experience shipping track 2 client libraries, we realized that the technical debt accumulated in `@azure/core-http` was costing us in terms of maintainability and performance. After multiple internal discussions about what exactly needed to change in order to move forward, a plan emerged to modernize `@azure/core-http` and give us a more robust core to build upon.

### Important goals and considerations:

- The request pipeline should be re-usable outside of AutoRest generated code (e.g. Cosmos.)
- The request pipeline should be easily customizable by client libraries and consumers.

> - Performance of the majority use case should be optimized and have few allocations.

Reviewer: few allocations of what? :)

Author: reworded this

> - Future improvements to codegen should be easy to phase in.

Reviewer: ```suggestion
- Future improvements to code generator should be easy to phase in.
```

or

```suggestion
- Future improvements to generated code should be easy to phase in.
```

## Scope of work

### Separation of Concerns

> One major goal of this work is to decompose the library into three separate packages: `@azure/core-https`, `@azure/core-client`, and `@azure/core-xml`. This is to enable smaller dependency graphs and bundles for individual client libraries.

Reviewer: decompose sounds bad.. I agree, this is just me being silly

```suggestion
One major goal of this work is to split the library into three separate packages: `@azure/core-https`, `@azure/core-client`, and `@azure/core-xml`. This is to enable smaller dependency graphs and bundles for individual client libraries.
```

> The first package, `@azure/core-https` will not be AutoRest-specific and is intended to be usable by any package that wants to make requests in a way that follows our SDK design guidelines.

Reviewer: ```suggestion
The first package, `@azure/core-https` will not be AutoRest-specific and is intended to be usable by any client that wants to make requests using a pipeline in a way that follows our SDK design guidelines.
```

`@azure/core-client` is intended to be a runtime dependency of any AutoRest-generated client library and has all necessary helpers except for XML (de)serialization. Packages that require XML support will additionally depend upon `@azure/core-xml`.

### Pipeline improvements

A major weakness of the existing JS Core was that the request pipeline system was difficult to customize at the library level. This placed an undue burden on authors of libraries and increased support costs. As other languages had better support for customizing the request pipeline, we had a desire to provide a simple and flexible way to add and order policies that could affect both HTTP requests and responses.

The [New Pipeline Design](https://github.com/Azure/azure-sdk-for-js/issues/8461) solves these concerns by providing a new abstraction that is able to reliably apply ordered policies with minimal configuration.

### Dropping legacy browser support

> As 2.0 is intended to be a modernization effort, it will **not** support Internet Explorer. This does mean that client libraries which need to support IE will not be able to upgrade until such time as they are not required to support legacy browsers.

Reviewer: A line on what we have to gain by dropping the support would be nice

Author: Added another line for this.

Azure customers are indirect consumers of this package via the SDKs for different Azure services. Therefore, a major version change would not be largely disruptive. Currently, in order to preserve compatibility and not complicate the story for servicing, a new package has been created that is currently named `@azure/core-https`. This name is not final.

## History, Context, and Motivations

In ancient times (~5 years ago), Microsoft provided SDKs for NodeJS developers wanting to interact with Azure services. Ultimately, many of these packages were generated using [AutoRest](https://github.com/Azure/Autorest) using a package called `autorest.nodejs`. The autogenerated code had the need for a common, re-usable core for performing HTTP operations, and thus [`ms-rest`](https://github.com/Azure/ms-rest-js) was created.

Another package `ms-rest-azure` was created to house Azure-specific components and authetication code that worked with a variety of credentials like Service Principal, Device Token, and Managed Identity. At this point, support for TypeScript came in the form of hand-authored type declaration files.

> Later, there emerged a desire for isomorphic packages, packages that worked in both the browser as well as Node. This required a new runtime and code generator. As part of this effort, code was generated in TypeScript directly and a new repo was created: `azure-sdk-for-js`.
> 
> A new code generator was created (`autorest.typescript`) that replaced the existing one (`autorest.nodejs`.) `ms-rest` became `@azure/ms-rest-js`. The authentication pieces from `ms-rest-azure` were put into `@azure/ms-rest-nodeauth` and an equivalent package was created for the browser called `@azure/ms-rest-browserauth`. All remaining parts of `ms-rest-azure` were placed into `@azure/ms-rest-azure-js`.

Reviewer: ```suggestion
Later, there emerged a desire for isomorphic packages, packages that worked in both the browser as well as Node. This required a new runtime, code generator and repo. As part of this effort, 
- `ms-rest` became `@azure/ms-rest-js` to provide the same capabilities in both Node.js and browser
- The authentication pieces from `ms-rest-azure` were put into `@azure/ms-rest-nodeauth` and an equivalent package was created for the browser called `@azure/ms-rest-browserauth`.
- All remaining parts of `ms-rest-azure` were placed into `@azure/ms-rest-azure-js`.
- The new code generator (`autorest.typescript`) that generated code in TypeScript and used the newer runtimes mentioned above replaced the existing one (`autorest.nodejs`) that generated JavaScript code.
- The packages thus generated got a new home in the repo `azure-sdk-for-js`. The older packages lived in `azure-sdk-for-node` repo.
```

> `@azure/ms-rest-js` was used as the basis for `@azure/core-http` when the JS team for the new Azure SDK was formed. This was nice as it bootstrapped the ability to make HTTP requests and provided broad compatibility with many track 1 clients that still exist in the `@azure/sdk-for-js` repository today.

Reviewer: ```suggestion
A year or so later, the new Azure SDK team was formed in Nov 2018 that then came up with guidelines on how to write an Azure SDK in TypeScript. To follow these guidelines, we needed yet another runtime. This was the new `@azure/core-http` package built over the existing  `@azure/ms-rest-js`. At the time we chose to do so as making slight changes and improvements to a tried and tested solution was preferred over investing time in a solution that had to be written from scratch.
```
